### ISSUE TITLE
provide Proxy Interface at the client side

### ISSUE THREAD

**Issue by nimo23** [eclipse-ee4j/jaxrs-api#769]
With RestEasy, we can define an interface to make restful calls at the client side:

```
@Path("https://www.service/query")
@Produces({ MediaType.APPLICATION_JSON, MediaType.TEXT_PLAIN })
public interface UserClient {
	
	@GET
	String getUser(@QueryParam("id") Integer id);

}
```

```
ResteasyWebTarget target = (ResteasyWebTarget) httpClient.target("");
var userClient = target.proxy(restClient);
// make GET-request and returns string of response body
userClient.getUser(1); 
```

Would be nice to provide this by jaxrs-api.

**Comment by jansupol** [comment]
This is basically part of the functionality of [Microprofile REST Client](https://download.eclipse.org/microprofile/microprofile-rest-client-1.3.2/?d). I assume the discussion about the relation between JAX-RS and the MP rest client will come

**Comment by nimo23** [comment]
In my opinion, **JAX-RS** should be merged with **Microprofile REST Client** and  should be called **"Open-EE-Restful"** with namespace "`openee.restful.*` or "`ee4j.restful`". The same goes for "Open-EE-Persistence", "Open-EE-Security", "Open-EE-Config", etc.

**Comment by ggam** [comment]
So you propose to abandon the Jakarta EE name?

**Comment by nimo23** [comment]
> So you propose to abandon the Jakarta EE name?

Imagine the follwing:

- Open-EE-RS (former: Microprofile REST Client + JAX-RS) with namespace `openee.rs.*`
- Open-EE-Persistence (former JPA) with namespace `openee.persistence.*`
- Open-EE-Security (former JEE-Security) with namespace `openee.security.*`
- Open-EE-Metrics (former Microprofile Health) with namespace `openee.metrics.*`
- etc.

I think, it is better to merge the efforts into Open-EE-X than continue with Jakarte EE and Microprofile side by side. In essence, Jakarta EE and Microprofile have the same goal: to provide enterprise features not available in Java SE. I think, both names will not be strong enough for the feature to make the intention clear and to attract. Forget the microservices and all the other buzzword which leads to microprofile..what is micro..why the island..what is "open enterprise edition" - you know, the last is clear: I **open** the **enterprise edition** for **restful** services with open.ee.restful.* (Open-EE-Restful).

**Comment by chkal** [comment]
I don't think that this issue is the best place to discuss your Open-EE idea. And I'm pretty sure that there won't be another name change.

**Comment by Cousjava** [comment]
@nimo23 See https://github.com/eclipse-ee4j/ee4j/issues/1 for the debate on naming. Open EE was considered and ultimately rejected in favour of Jakarta EE.

**Comment by nimo23** [comment]
@Cousjava thanks for sharing the link. I did not know about it. 

The relation between JAX-RS and the MP rest client (and in general, the microprofile and jakarte ee) must come because it's a pitty to spend resources on projects with so much overlapping instead of merging the efforts into one.

**Comment by derekm** [comment]
This is also something I've been using since 2012 as Jersey's `WebResourceFactory`:

https://github.com/eclipse-ee4j/jersey/blob/master/ext/proxy-client/src/main/java/org/glassfish/jersey/client/proxy/WebResourceFactory.java

When MicroProfile REST Client spec was being written, I made sure the spec writers considered `@BeanParam`s and sub-resource locators, etc.

I have many facilities built around `WebResourceFactory`.

**Comment by derekm** [comment]
Some interesting PRs involving `WebResourceFactory` proxy client from before the Eclipse donation:

JERSEY-1653: Add support for @BeanParam in jersey-proxy-client
https://github.com/jersey/jersey/pull/235

JERSEY-3248: support parameter converters in WebResourceFactory
https://github.com/jersey/jersey/pull/269
